### ISSUE TITLE
Cannot update offline map files

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#11184]
A beta users reports, that c:geo crashes when he follows the update notification for the offline map.

## c:geo version 

2021.07.10-RC

## System information 

```
## System information

c:geo version: 2021.07.10-RC

Device:
-------
- Device type: SM-G960F (starlteeea, samsung)
- Available processors: 8
- Android version: 10
- Android build: QP1A.190711.020.G960FXXUFFUE1
- Screen resolution: 1080x2076px (360x668dp)
- Sailfish OS detected: false
- Google Play services: enabled - 21.21.16 (120408-378233385)
- HW acceleration: enabled (default state)

Sensor and location:
-------
- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:
-------
- Hide caches: foundarchived
- Hide waypoints: visited
- Set language: de
- System date format: dd.MM.yy
- Debug mode active: no
- Live map mode: false
- OSM multi-threading: true / threads: 4
- Global filter: display all caches
- Last backup: 11. Juli, 13:47
- Routing mode: Off
- Settings: v5, Count:158
- Map: Germany (Offline)
- Id: cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider$OfflineMapSource:3735-3561:cgeo/Maps/germany.map
- Atts: Map data (c) OpenStreetMap contributors
- Theme: fzk-outdoor-soft-v5.zip:fzk-outdoor-soft-v5/fzk-outdoor-soft-v5.xml

Services:
-------
- Geocaching sites enabled:
geocaching.com: Logged in (Anmeldung OK) / PREMIUM
Geocaching.com Adventure Lab
- Geocaching.com date format: MM/dd/yyyy
- Routing: external / BRouter installed: true
- Installed c:geo plugins: contacts

Permissions & paths:
-------
- Fine location permission: granted
- Write external storage permission: granted
- System internal c:geo dir: /data/user/0/cgeo.geocaching (15,9 GB free) v2 internal isDir(8 entries)
- Legacy User storage c:geo dir: /storage/emulated/0/cgeo (15,9 GB free) v2 external non-removable isDir(7 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (15,9 GB free) v2 external non-removable isDir(2924 entries)
- Internal theme sync (is turned off): /data/user/0/cgeo.geocaching/MapThemeData (15,9 GB free) v2 internal isDir(0 entries)
- Public Folders: #11
- BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=7, dirs:>=8, totalFileSize:>=81,2 KB, free space: 15,9 GB, files on device: 1740800)
- OFFLINE_MAPS: SD-Karte/cgeo/Maps (User-Defined)[SD-Karte/cgeo/Maps[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/3735-3561%3Acgeo%2FMaps::]] (Uri: content://com.android.externalstorage.documents/tree/3735-3561%3Acgeo%2FMaps/document/3735-3561%3Acgeo%2FMaps, Av:true, files:5, dirs:1, totalFileSize:3,2 GB, free space: 89,1 GB, files on device: 0)
- OFFLINE_MAP_THEMES: SD-Karte/cgeo/Themes (User-Defined)[SD-Karte/cgeo/Themes[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/3735-3561%3Acgeo%2FThemes::]] (Uri: content://com.android.externalstorage.documents/tree/3735-3561%3Acgeo%2FThemes/document/3735-3561%3Acgeo%2FThemes, Av:true, files:6, dirs:0, totalFileSize:1,2 MB, free space: 89,1 GB, files on device: 0)
- LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:5, dirs:0, totalFileSize:29,4 MB, free space: 15,9 GB, files on device: 1740800)
- GPX: Downloads:msd/19805 (User-Defined)[Downloads:msd/19805[DOCUMENT#0:p-content://com.android.providers.downloads.documents/tree/msd%3A19805::]] (Uri: content://com.android.providers.downloads.documents/tree/msd%3A19805/document/msd%3A19805, Av:true, files:5, dirs:0, totalFileSize:110,2 MB, free space: 15,9 GB, files on device: 1740800)
- BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:6, dirs:3, totalFileSize:525,7 MB, free space: 15,9 GB, files on device: 1740800)
- FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 15,9 GB, files on device: 1740800)
- SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 15,9 GB, files on device: 1740800)
- ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:7, dirs:1, totalFileSize:81,2 KB, free space: 15,9 GB, files on device: 1740800)
- ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 15,9 GB, files on device: 1740800)
- TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 15,9 GB, files on device: -1)
- Map render theme path: fzk-outdoor-soft-v5.zip:fzk-outdoor-soft-v5/fzk-outdoor-soft-v5.xml
- PersistedDocumentUris: #1
- TRACK: null
- Persisted Uri Permissions: #4
- content://com.android.providers.downloads.documents/tree/msd%3A19805 (19. Apr., 21:44):RW
- content://com.android.externalstorage.documents/tree/3735-3561%3Acgeo%2FThemes (27. Apr., 00:24):RW
- content://com.android.externalstorage.documents/tree/primary%3Acgeo (21. März, 15:58):RW
- content://com.android.externalstorage.documents/tree/3735-3561%3Acgeo%2FMaps (26. Apr., 23:41):RW
- Database: /data/user/0/cgeo.geocaching/databases/data (v95, Size:201,5 MB) on system internal storage

--- End of system information ---
```

## Additional context 

Log file has this:
```
07-11 13:47:58.357 21348 21348 E AndroidRuntime: FATAL EXCEPTION: main
07-11 13:47:58.357 21348 21348 E AndroidRuntime: Process: cgeo.geocaching, PID: 21348
07-11 13:47:58.357 21348 21348 E AndroidRuntime: android.os.NetworkOnMainThreadException
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:2098)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2056)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:188)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.content.ContentProviderProxy.insert(ContentProviderNative.java:481)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.content.ContentResolver.insert(ContentResolver.java:1835)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.app.DownloadManager.enqueue(DownloadManager.java:1544)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at cgeo.geocaching.downloader.DownloaderUtils$CheckForDownloadsTask.lambda$onPostExecuteInternal$0(DownloaderUtils.java:303)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at cgeo.geocaching.downloader.-$$Lambda$DownloaderUtils$CheckForDownloadsTask$FLWPWkG6wcn0q3pCXeEgr9guA1o.onClick(Unknown Source:6)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at androidx.appcompat.app.AlertController$ButtonHandler.handleMessage(AlertController.java:167)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:237)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8167)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
07-11 13:47:58.357 21348 21348 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
```

**Comment by Lineflyer** [comment]
It happens after confirming the following dialog:
![image](https://user-images.githubusercontent.com/949669/125210697-bb01c700-e2a1-11eb-9e81-df3c63fd3d3c.png)


Note also the non-scrollable(?) text....thats another issue.

**Comment by fm-sys** [comment]
This seem to be the affected line - somehow the toast can't be displayed:

https://github.com/cgeo/cgeo/blob/41a87ca6c0e95703bd4331fe686702d517f75ebb/main/src/cgeo/geocaching/downloader/DownloaderUtils.java#L303

**Comment by eddiemuc** [comment]
Problem is the call to `DownloadManager.enqueue` which apparently is not allowed to happen in the main GUI thread. This call must be moved to a background task.

**Comment by fm-sys** [comment]
That sounds way more logic. I was already wondering how a Toast can cause a `NetworkOnMainThreadException` 😅 

Well, I'm still learning how to read these confusing stack traces correctly... 🙈

**Comment by eddiemuc** [comment]
> Well, I'm still learning how to read these confusing stack traces correctly... 🙈

Problem is you can't really trust the line numbers because we are editing the code so much lately... While the call to `DownloadManager.enqueue` obviously was in line 303 at the time the stack trace was logged, recent edits have obviously moved this line to 301.

**Comment by fm-sys** [comment]
Unfortunately, I will not be able to work on a fix, as I don't have Android Studio currently at hand. 

So either somebody else needs step in, or we have to wait some days for the fix...

**Comment by moving-bits** [comment]
According to both [documentation](https://developer.android.com/reference/android/app/DownloadManager) and several SO articles I don't see any reason why `enqueue` should not be called on the main thread, as `enqueue` only adds an entry to the download queue. The actual download is done in the background anyway - that's what the `DownloadManager` is for. - But it may be some version- or vendor-specific problem.

Additionally, I found one SO article mentioning a similar problem, but happening only on Samsung devices. Suggested solution is to change the download folder by using `.setDestinationInExternalFilesDir(context, Environment.DIRECTORY_DOWNLOADS, fileName)` instead of `.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, fileName)` - which would fit to our case, as we are using `Environment.DIRECTORY_DOWNLOADS` as well. - I don't see why this should be related to network traffic, but for the user it seems to have worked.

**Comment by moving-bits** [comment]
As discussed between @Lineflyer and me offline: Let's test the possible fix #11393 on `master` first for a while, and thus remove it from the milestone list for the next release. If it works, we can still cherry-pick it for a small bugfix-update.

**Comment by moving-bits** [comment]
Ok, I could do with some Java advice. To put all calls do `DownloadManager.enqueue()` into the background, I need to solve the following in one situation:

```
private int addToDownload(...) {
    int id = DownloadManager.enqueue(...);
    return id;
}
```

How can I put the call to `enqueue` into the background, but return from that function only after having received the id returned by the call?

Or otherwise, since it's only in one place (the list of maps found): How can I enqueue in the background, but trigger some update to the list of available downloads on having finished enqueing? I know there's something like Observers, but I'm not really firm with it, and even don't know, if it does fit here.
Any help appreciated.

**Comment by moving-bits** [comment]
Provided a fix putting `DownloadManager.enqueue` into background for the update check, as all error reports on Play Console are mentioning only that call. Let's see if that helps with the issue here.

**Comment by Lineflyer** [comment]
@moving-bits 
A user reported the same problem for 2021.12.24. I assume your change mentioned above is only on `master` thus not yet contained in the release?

```
12-24 18:32:12.108 30352 30352 E AndroidRuntime: FATAL EXCEPTION: main
12-24 18:32:12.108 30352 30352 E AndroidRuntime: Process: cgeo.geocaching, PID: 30352
12-24 18:32:12.108 30352 30352 E AndroidRuntime: android.os.NetworkOnMainThreadException
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:2098)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2056)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:188)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.content.ContentProviderProxy.insert(ContentProviderNative.java:481)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.content.ContentResolver.insert(ContentResolver.java:1842)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.app.DownloadManager.enqueue(DownloadManager.java:1544)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at cgeo.geocaching.downloader.DownloaderUtils$CheckForDownloadsTask.lambda$onPostExecuteInternal$0(DownloaderUtils.java:294)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at cgeo.geocaching.downloader.-$$Lambda$DownloaderUtils$CheckForDownloadsTask$FLWPWkG6wcn0q3pCXeEgr9guA1o.onClick(Unknown Source:6)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at androidx.appcompat.app.AlertController$ButtonHandler.handleMessage(AlertController.java:167)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:237)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8167)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
12-24 18:32:12.108 30352 30352 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: FATAL EXCEPTION: main
12-24 18:32:42.438 30720 30720 E AndroidRuntime: Process: cgeo.geocaching, PID: 30720
12-24 18:32:42.438 30720 30720 E AndroidRuntime: android.os.NetworkOnMainThreadException
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:2098)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2056)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:188)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.content.ContentProviderProxy.insert(ContentProviderNative.java:481)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.content.ContentResolver.insert(ContentResolver.java:1842)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.app.DownloadManager.enqueue(DownloadManager.java:1544)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at cgeo.geocaching.downloader.DownloaderUtils$CheckForDownloadsTask.lambda$onPostExecuteInternal$0(DownloaderUtils.java:294)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at cgeo.geocaching.downloader.-$$Lambda$DownloaderUtils$CheckForDownloadsTask$FLWPWkG6wcn0q3pCXeEgr9guA1o.onClick(Unknown Source:6)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at androidx.appcompat.app.AlertController$ButtonHandler.handleMessage(AlertController.java:167)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:237)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8167)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
12-24 18:32:42.438 30720 30720 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: FATAL EXCEPTION: main
01-01 15:57:05.991 32546 32546 E AndroidRuntime: Process: cgeo.geocaching, PID: 32546
01-01 15:57:05.991 32546 32546 E AndroidRuntime: android.os.NetworkOnMainThreadException
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:2098)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2056)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:188)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.content.ContentProviderProxy.insert(ContentProviderNative.java:481)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.content.ContentResolver.insert(ContentResolver.java:1842)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.app.DownloadManager.enqueue(DownloadManager.java:1544)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at cgeo.geocaching.downloader.DownloaderUtils$CheckForDownloadsTask.lambda$onPostExecuteInternal$0(DownloaderUtils.java:294)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at cgeo.geocaching.downloader.-$$Lambda$DownloaderUtils$CheckForDownloadsTask$FLWPWkG6wcn0q3pCXeEgr9guA1o.onClick(Unknown Source:6)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at androidx.appcompat.app.AlertController$ButtonHandler.handleMessage(AlertController.java:167)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:237)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8167)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
01-01 15:57:05.991 32546 32546 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
```

**Comment by moving-bits** [comment]
> @moving-bits A user reported the same problem for 2021.12.24. I assume your change mentioned above is only on `master` thus not yet contained in the release?

Yes, that's #12137 and merged to `master` only. Shall I cherry-pick and provide a PR for `release`?
(Might create merge conflicts, though, as further changes have moved the specific line to a totally different place meanwhile.)

**Comment by Lineflyer** [comment]
Seemingly a rather rare issue, otherwise we would have more reports (and we could reproduce). Lets just wait for the next feature release.

**Comment by Lineflyer** [comment]
@moving-bits Can we assume this fixed? At least no such reports anymore on support mail and there was a related change you mentioned above.

**Comment by moving-bits** [comment]
Ok, then let's close here.
